src/VisualStudio/Core/Def/Implementation/Remote/RemotableDataJsonRpcEx.cs

```diff
@@ -69,60 +69,42 @@ public async Task RequestAssetAsync(int scopeId, Checksum[] checksums, string st
         {
             try
             {
-                using (var combinedCancellationToken = _shutdownCancellationSource.Token.CombineWith(cancellationToken))
+                using var combinedCancellationToken = _shutdownCancellationSource.Token.CombineWith(cancellationToken);
+
                 using (Logger.LogBlock(FunctionId.JsonRpcSession_RequestAssetAsync, streamName, combinedCancellationToken.Token))
-                using (var stream = await DirectStream.GetAsync(streamName, combinedCancellationToken.Token).ConfigureAwait(false))
                 {
-                    using (var writer = new ObjectWriter(stream, combinedCancellationToken.Token))
-                    {
-                        writer.WriteInt32(scopeId);
-
-                        await WriteAssetAsync(writer, scopeId, checksums, combinedCancellationToken.Token).ConfigureAwait(false);
-                    }
-
-                    await stream.FlushAsync(combinedCancellationToken.Token).ConfigureAwait(false);
+                    await ClientDirectStream.WriteDataAsync(
+                        streamName,
+                        (scopeId, checksums),
+                        (writer, data, ct) => WriteAssetAsync(writer, data.scopeId, data.checksums, ct),
+                        combinedCancellationToken.Token).ConfigureAwait(false);
                 }
             }
-            catch (Exception ex) when (ReportUnlessCanceled(ex, cancellationToken))
+            catch (Exception ex) when (_endPoint.ReportAndPropagateUnexpectedException(ex, cancellationToken))
             {
-                // only expected exception will be catched. otherwise, NFW and let it propagate
-                Debug.Assert(cancellationToken.IsCancellationRequested || ex is IOException);
+                throw ExceptionUtilities.Unreachable;
```

Comment:
more cases of not understanding why this is hte right exception pattern.

Reply:
A comment on `ReportAndPropagateUnexpectedException` explains a bit. In general what we want is to intercept and report NFW for any unexpected exception on the process boundary, so that we have good memory dump to investigate.  The exception should however be just propagated so that it flows thru the RPC call to the other side and is rethrown there as `RemoteInvocationException`. 

In this case `RequestAssetAsync` is method invoked remotely, so we want to do the intercepting before it returns.

Reply:
What's uncelar to me is why tehre seems to be so much exception handling all over the place.  i guess i would have expected it in two place (on the host side at least):

1. when making connections
2. when making remote calls.

it *feels* (though maybe i'm wrong) that there are far more places than that, and they all tend to do things their own way.

Reply:
That's essentially the goal of this PR :). There is one more place where we want to intercept exceptions - when we leave remote call. There we keep the exception going, but record NFW so we don't loose its original stack.

Reply:
If you look at the final version of the PR you'll find all exception handling in `RemoteEndPoint`. If you see it elsewhere let me know, I might have missed something.